### PR TITLE
Docs: Add resource handler topic

### PR DESCRIPTION
**Pull request by josmperez**
Convert [community forum post ](https://community.grafana.com/t/how-to-add-a-resource-handler-for-your-data-source/58924)into draft topic for "Add resource handler to data source plugins".

Fixes #

[Issue 66401](https://github.com/grafana/grafana/issues/66401)

### CODE REVIEW

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+:::
+
+### Additional examples
+
+Some other examples of using the [`httpadapter`](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter) package can be found for some of the built-in Grafana data sources:
+
+- Test data source [grafana/testdata.go](https://github.com/grafana/grafana/blob/30aa24a18396f61bfbcbfbeee78b1c0d23685fc3/pkg/tsdb/testdatasource/testdata.go#L41-L42) at 30aa24a18396f61bfbcbfbeee78b1c0d23685fc3 - See code at the [grafana/grafana](https://github.com/grafana/grafana/blob/30aa24a18396f61bfbcbfbeee78b1c0d23685fc3/pkg/tsdb/testdatasource/testdata.go#L41-L42) repo.
+- Test data source [grafana/resource_handler.go](https://github.com/grafana/grafana/blob/30aa24a18396f61bfbcbfbeee78b1c0d23685fc3/pkg/tsdb/testdatasource/resource_handler.go#L18) at 30aa24a18396f61bfbcbfbeee78b1c0d23685fc3 - See code at the [grafana/grafana](https://github.com/grafana/grafana/blob/30aa24a18396f61bfbcbfbeee78b1c0d23685fc3/pkg/tsdb/testdatasource/resource_handler.go#L18) repo.
+
+### What if you need access to the backend plugin context?
```
Comment (marefr) [inline docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:163]: I would suggest this section to go before the examples section.

**Review by marefr — CHANGES_REQUESTED** [review]
First of all, sorry for keeping you waiting. This looks pretty good, added some suggestions and updates since things have changed a bit since that community post was written.

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  - datasource
+  - resource
+  - resource handler
+---
+
+# Add resource handler for data source plugins
+
+You can add a resource handler to your data source to extend the Grafana HTTP API with your own data source-specific routes. This guide explains why you may want to add resource handlers and some common ways for doing so.
+
+## Uses of resource handlers
```
[the remark this answers is not in the document]
Comment (josmperez) [reply docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:19]: added this link, good idea.

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
+---
+id: add-resource-handler
+title: Add resource handler for data source plugins
```
[the remark this answers is not in the document]
Comment (josmperez) [reply docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:3]: I've captured this suggestion here: https://github.com/grafana/plugin-tools/issues/878

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+	_, err := rw.Write(bytes)
+	if err != nil {
+		return
+	}
+	rw.WriteHeader(http.StatusOK)
+}
+```
+
+Adding resource handlers to your backend plugins opens up more ways to make your plugin more dynamic.
```
[the remark this answers is not in the document]
Comment (josmperez) [reply docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:184]: agreed. removed it.

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md (hunk 1 above)
Comment (josmperez) [reply docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:165]: done.

**Review by marefr — CHANGES_REQUESTED** [review]
Changes look good 👍 Still some unresolved comments. Let me know if you need help with the removal of fetch example

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		})
+	default:
+		return sender.Send(&backend.CallResourceResponse{
+			Status: http.StatusNotFound,
+		})
+	}
+}
+```
+
+With this, your plugin now has its own REST API that you can query from your query editor, using `BackendSrv.fetch()`.
```
[the remark this answers is not in the document]
Comment (josmperez) [reply docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md]: done.

**Comment by josmperez** [comment]
> Changes look good 👍 Still some unresolved comments. Let me know if you need help with the removal of fetch example

I think I got the unresolved comments resolved now.

**Review by marefr — APPROVED** [review]
LGTM

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+}
+```
+
+You can then access your resources through the following endpoint: `http://<grafana instance host:port>/api/datasources/uid/<DATASOURCE_IUD>/resources`
+
+In this example code, `DATASOURCE_UID` is the data source unique identifier (UID) that uniquely identifies your data source.
+
+:::tip
+
+To verify the data source UID, you can enter `window.grafanaBootData.settings.datasources` in the Browser Developer Tools Console, to list all the configured data sources in your Grafana instance.
```
Comment (sympatheticmoose) [inline docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:44]:
```suggestion
To verify the data source UID, you can enter `window.grafanaBootData.settings.datasources` in your browser's developer tools console, to list all the configured data sources in your Grafana instance.
```

Not sure all the capitals are necessary?

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+```
+const namespaces = await props.datasource.getResource('namespaces');
+props.datasource.postResource('device', { state: "on" });
+```
+
+## Advanced use cases
+
+If you have some more advanced use cases or want to use a more Go-agnostic approach for handling resources, you can use the regular [`http.Handler`](https://pkg.go.dev/net/http#Handler). You can do so by using a package provided by the [Grafana Plugin SDK for Go](../../introduction/grafana-plugin-sdk-for-go.md) named [`httpadapter`](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter). This package provides support for handling resource calls using an [`http.Handler`](https://pkg.go.dev/net/http#Handler).
+
+What’s interesting with using [`http.Handler`](https://pkg.go.dev/net/http#Handler) is that you can use it with Go’s builtin router functionality called [`ServeMux`](https://pkg.go.dev/net/http#ServeMux) or use your preferred HTTP router library (for example, [`gorilla/mux`](https://github.com/gorilla/mux)).
```
Comment (sympatheticmoose) [inline docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:86]:
```suggestion
What’s interesting with using [`http.Handler`](https://pkg.go.dev/net/http#Handler) is that you can use it with Go’s built-in router functionality called [`ServeMux`](https://pkg.go.dev/net/http#ServeMux) or use your preferred HTTP router library (for example, [`gorilla/mux`](https://github.com/gorilla/mux)).
```

Comment (sympatheticmoose) [reply docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:86]: "What's interesting" seems more conversational than usual and we don't really explain why its useful to use ServeMux or other libraries. Not a blocker, just a suggestion
```suggestion
Using [`http.Handler`](https://pkg.go.dev/net/http#Handler) allows you to also use Go’s built-in router functionality called [`ServeMux`](https://pkg.go.dev/net/http#ServeMux) or your preferred HTTP router library (for example, [`gorilla/mux`](https://github.com/gorilla/mux)).
```

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+An alternative to using the [`CallResource` method shown in the above example](#add-resource-handler-for-data-source-plugins) is to use [`httpadapter`](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter) and [`ServeMux`](https://pkg.go.dev/net/http#ServeMux) as shown below:
```
Comment (sympatheticmoose) [inline docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:88]:
```suggestion
An alternative to using the `CallResource` method shown in the [above example](#implement-the-resource-handler-interface) is to use [`httpadapter`](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter) and [`ServeMux`](https://pkg.go.dev/net/http#ServeMux) as shown below:
```

Link didn't work and having code blocks in links still looks bad.

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	_, err := rw.Write(bytes)
+	if err != nil {
+		return
+	}
+	rw.WriteHeader(http.StatusOK)
+}
+```
+### Additional examples
```
Comment (sympatheticmoose) [inline docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:161]:
```suggestion
## Additional examples
```

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+- Built-in TestData datasource, [create resource handler](https://github.com/grafana/grafana/blob/5687243d0b3bad06c4da809f925cfdf3d32c5a16/pkg/tsdb/grafana-testdata-datasource/testdata.go#L45) and [register routes](https://github.com/grafana/grafana/blob/5687243d0b3bad06c4da809f925cfdf3d32c5a16/pkg/tsdb/grafana-testdata-datasource/resource_handler.go#L17-L28).
```
Comment (sympatheticmoose) [inline docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:168]:
```suggestion
- Grafana's built-in TestData datasource, [create resource handler](https://github.com/grafana/grafana/blob/5687243d0b3bad06c4da809f925cfdf3d32c5a16/pkg/tsdb/grafana-testdata-datasource/testdata.go#L45) and [register routes](https://github.com/grafana/grafana/blob/5687243d0b3bad06c4da809f925cfdf3d32c5a16/pkg/tsdb/grafana-testdata-datasource/resource_handler.go#L17-L28).
```

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+		return
+	}
+	rw.WriteHeader(http.StatusOK)
+}
+```
+### Additional examples
+
+Some other examples of using resource handlers and the [`httpadapter`](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter) package:
+
+- [datasource-basic](https://github.com/grafana/grafana-plugin-examples/tree/main/examples/datasource-basic) example:
```
Comment (sympatheticmoose) [inline docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:165]:
```suggestion
- The [datasource-basic](https://github.com/grafana/grafana-plugin-examples/tree/main/examples/datasource-basic) example:
```

**Review by sympatheticmoose — COMMENTED** [review]
Some mostly style suggestions, happy trusting Marcus for technical review 😄

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+  - plugin
+  - data source
+  - datasource
+  - resource
+  - resource handler
+---
+
+# Add resource handler for data source plugins
+
+You can add a resource handler to your data source to extend the Grafana HTTP API with your own data source-specific routes. This guide explains why you may want to add [resource](/introduction/backend-plugins#resources) handlers and some common ways for doing so.
```
[the remark this answers is not in the document]
Comment (marefr) [reply docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:17]: I think this one is the last outstanding issue

[the remark this answers is not in the document]
Comment (marefr) [reply docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:17]: Optimal to change the title but it becomes very long so suggest wait with that until following up the related issue for app plugins. 

```suggestion
You can add a resource handler to your data source backend to extend the Grafana HTTP API with your own data source-specific routes. This guide explains why you may want to add [resource](/introduction/backend-plugins#resources) handlers and some common ways for doing so.
```

**Review by marefr — APPROVED** [review]
LGTM with my last suggestion

File: docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
+---
+id: add-resource-handler
+title: Add resource handler for data source plugins
+description: Add a resource handler for data source plugins.
```
Comment (josmperez) [inline docusaurus/docs/create-a-plugin/extend-a-plugin/add-resource-handler.md:4]:
```suggestion
description: Learn how to add a resource handler for data source plugins.
```

**Pull request merged**